crates/bevy_picking/src/mesh_picking/ray_cast/intersections.rs

```diff
@@ -0,0 +1,405 @@
+    tri_vertices: [Vec3A; 3],
+    tri_normals: Option<[Vec3A; 3]>,
+    max_distance: f32,
+    ray: &Ray3d,
+    backface_culling: Backfaces,
+) -> Option<RayMeshHit> {
+    let hit = ray_triangle_intersection(ray, &tri_vertices, backface_culling)?;
+
+    if hit.distance < 0.0 || hit.distance > max_distance {
+        return None;
+    };
+
+    let point = ray.get_point(hit.distance);
+    let u = hit.barycentric_coords.0;
+    let v = hit.barycentric_coords.1;
+    let w = 1.0 - u - v;
+    let barycentric = Vec3::new(u, v, w);
+
+    let normal = if let Some(normals) = tri_normals {
+        normals[1] * u + normals[2] * v + normals[0] * w
+    } else {
+        (tri_vertices[1] - tri_vertices[0])
+            .cross(tri_vertices[2] - tri_vertices[0])
+            .normalize()
+    };
+
+    Some(RayMeshHit {
+        point,
+        normal: normal.into(),
+        barycentric_coords: barycentric,
+        distance: hit.distance,
+        triangle: Some(tri_vertices),
+        triangle_index: None,
+    })
+}
+
+/// Takes a ray and triangle and computes the intersection.
+#[inline(always)]
+fn ray_triangle_intersection(
+    ray: &Ray3d,
+    triangle: &[Vec3A; 3],
+    backface_culling: Backfaces,
+) -> Option<RayTriangleHit> {
+    // Source: https://www.scratchapixel.com/lessons/3d-basic-rendering/ray-tracing-rendering-a-triangle/moller-trumbore-ray-triangle-intersection
+    let vector_v0_to_v1: Vec3A = triangle[1] - triangle[0];
+    let vector_v0_to_v2: Vec3A = triangle[2] - triangle[0];
+    let p_vec: Vec3A = (Vec3A::from(*ray.direction)).cross(vector_v0_to_v2);
+    let determinant: f32 = vector_v0_to_v1.dot(p_vec);
+
+    match backface_culling {
+        Backfaces::Cull => {
+            // if the determinant is negative the triangle is back facing
+            // if the determinant is close to 0, the ray misses the triangle
+            // This test checks both cases
+            if determinant < f32::EPSILON {
+                return None;
+            }
+        }
+        Backfaces::Include => {
+            // ray and triangle are parallel if det is close to 0
+            if determinant.abs() < f32::EPSILON {
+                return None;
+            }
+        }
+    }
+
+    let determinant_inverse = 1.0 / determinant;
+
+    let t_vec = Vec3A::from(ray.origin) - triangle[0];
+    let u = t_vec.dot(p_vec) * determinant_inverse;
+    if !(0.0..=1.0).contains(&u) {
+        return None;
+    }
+
+    let q_vec = t_vec.cross(vector_v0_to_v1);
+    let v = Vec3A::from(*ray.direction).dot(q_vec) * determinant_inverse;
+    if v < 0.0 || u + v > 1.0 {
+        return None;
+    }
+
+    // The distance between ray origin and intersection is t.
+    let t: f32 = vector_v0_to_v2.dot(q_vec) * determinant_inverse;
+
+    Some(RayTriangleHit {
+        distance: t,
+        barycentric_coords: (u, v),
+    })
+}
+
+// TODO: It'd be nice to use `RayCast3d` from `bevy_math` instead. It caches the direction reciprocal.
```

Comment:
Is there a reason this `TODO` hasn't been acted on?

Reply:
I realized it probably wouldn't work: you can't use the cached reciprocal because the ray needs to be transformed into model space first. We also can't reuse `RayCast3d::aabb_intersection_at` because it assumes normalized rays, while this method (I think) can have scaled rays because of the transformation.

I changed this implementation now to be otherwise closer to `RayCast3d::aabb_intersection_at` though, since it takes advantage of SIMD better and is a bit faster.

Reply:
Ok that makes sense, thanks for clarifying!